Refactor region import export code

pford opened the pull request:
**Description**

* What is implemented or fixed? Mention the linked issue(s), if any.
Closes #1505 

* How does this PR solve the issue? Give a brief summary.

  - Separated import/export code, very little overlap between the two:
    - RegionImportExport -> RegionImporter, RegionExporter (base classes)
    - CrtfImportExport -> CrtfImporter, CrtfExporter
    - Ds9ImportExport -> Ds9Importer, Ds9Exporter
   - Moved code common to CRTF/DS9 import/export to RegionImportExportUtil
   - Moved code where possible from RegionHandler, such as adding getters for region file-specific importer and exporter to RegionImportExport.h
   - Since there were so many files, created src/Region/RegionImportExport subdirectory
   - Pruned unnecessary includes and added docstrings to header files in this subdirectory
   - Did not change implementation

* Are there any companion PRs (frontend, protobuf)?
No, backend code maintenance only

* Is there anything else that testers should know (e.g. exactly how to reproduce the issue)?
All region import/export tests should pass

**Checklist**

- [x] changelog updated / ~no changelog update needed~
- [ ] e2e test passing / corresponding fix added / new e2e test created
- [ ] ICD test passing / corresponding fix added / new ICD test created
- [x] ~protobuf updated to the latest dev commit~ / no protobuf update needed
- [x] ~protobuf version bumped~ / protobuf version not bumped
- [x] added reviewers and assignee
- [ ] GitHub Project estimate added

pford committed:
Refactor region import and export into separate files in new dir

pford committed:
Rename Import->Importer and Export->Exporter and add docstrings

confluence reviewed (commented):
This all looks sensible -- there are just a few doxygen errors caused by parameters being renamed.

pford committed:
Fix missing docstrings

confluence reviewed (changes requested):
1. I suggest renaming some things for clarity:

    * `_export_regions` is ambiguous ("regions for export" vs "export (verb) regions", especially since there's also a function with the same name and the other meaning). How about `_region_lines`?
    * Similarly, I would eliminate use of "export" as an adjective from all the function names. `AddExportRegion` can just be `AddRegion` (the way this is accessed, i.e. `exporter->AddRegion`, makes it unambiguous). `AddRegionExportLine` and `AddTextRegionExportLines` can also have the `Export` removed. `AddExportRegionPixel` and `AddExportRegionWorld` return a region line and don't add anything, so they should be renamed to something like `GetRegionLinePixel` and `GetRegionLineWorld`.
    * Similarly, I would rename `_import_regions` to just `_regions` (I think `_import_errors` is OK, but could also be renamed to `_errors`). `GetImportedRegions` can just be `GetRegions`.
    * For consistency, I suggest renaming `SetRegion` in the importer classes to `ImportRegion`.
    * This still leaves asymmetry in the naming conventions in the two directions (in the export, convert = "add" and write out = "export", and in the import, read in = "read" and convert = "import"), but at least they'll be consistent within each component, so this isn't that urgent and I'm happy to leave it for now.

2. After this, some of the docstrings could be rephrased more clearly, e.g. for the first `AddRegion` in `RegionExporter`: "Export region in pixel coordinates using RegionState and RegionStyle to _export_regions." -> something like "Serialise region provided in pixel coordinates to appropriate file format and add to vector to be exported." (I don't think it's necessary to mention the parameters, which are described separately) and "Whether the region export is successful" -> "Whether the region was added successfully"; and for `ExportRegions`: "Export region file lines in _export_regions to file." -> something like "Export serialised regions accumulated in vector to file." (Or "serialize"; I have no strong opinion! 🙂)

3. General docstring issue: if additional docstrings are provided in the `.cc` file, they should be `@details` or `@note` entries with more detail about the implementation. They should not be copies of the corresponding object docstrings in `.h`. In the generated documentation, both docstrings are merged. (See the Stokes utils for an example.)\
\
The copies in e.g. the exporter files are merged correctly (not duplicated), but maintaining identical copies in two places is unnecessarily complicated. I suggest removing the copies from the `.cc` files, but moving any information that refers to implementation internals (e.g. internal variable names) from the `.h` docstrings to `.cc`.\
\
I think that references to internal variable names can be avoided entirely if the docstrings are reworded, but wherever something *should* be mentioned about internals, it should go in the `.cc` docstring.

pford committed:
Remove Import/Export from variable names and improve docstrings

pford committed:
Fix parameter names in docstrings

confluence commented on src/Region/RegionImportExport/CrtfExporter.cc:
```diff
@@ -464,8 +451,7 @@ char CrtfExporter::GetAnnSymbolCharacter(CARTA::PointAnnotationShape point_shape
 
-std::string CrtfExporter::GetRegionColor(const CARTA::RegionStyle& region_style) {
-    // Remove beginning # and convert to lower case from color string in region style.
+std::string CrtfExporter::GetStyleColor(const CARTA::RegionStyle& region_style) {
```
I would just call this `GetColor` for consistency with the other names.

pford replied:
`FormatColor()` in RegionImportExportUtils is used by CrtfImporter, Ds9Importer, and Ds9Exporter.  I changed this name to `FormatColor` with a string parameter for consistency.

confluence reviewed (changes requested):
This is great, thanks! I think the names and docstrings are much clearer now. I only have one function name suggestion.

Please disregard my previous comments about duplicate docstrings -- I don't know what I was looking at! I think that at some point I must have seen two similar `.h` files in sequence because a large `.cc` diff was being hidden. 🙈

pford committed:
Change name of method to format color string

kswang1029 reviewed (changes requested):
@pford e2e tests picked up a few issues

1) in the exported ds9 region file, the reference is changed from fk5 to J2000.
Now it is 
```
# Region file format: DS9 CARTA 6.0.0-dev
global color=green dashlist=8 3 width=1 font="helvetica 10 normal roman" select=1 highlite=1 dash=0 fixed=0 edit=1 move=1 delete=1 include=1 source=1
J2000
point(269.090232042, -21.955048017) # color=#2EE6D6 width=2
line(269.087708606, -21.955757242, 269.089696784, -21.957175681) # color=#2EE6D6 width=2
box(269.088702690, -21.956466464, 6.1277", 8.1702", 0) # color=#2EE6D6 width=2
ellipse(269.088702690, -21.956466464, 2.5532", 5.1064", 90) # color=#2EE6D6 width=2
polygon(269.087555672, -21.955402630, 269.087937997, -21.958168590, 269.088702690, -21.956466464) # color=#2EE6D6 width=2
polyline(269.088702690, -21.957884904, 269.089926194, -21.957601211, 269.087479189, -21.957317524, 269.088702690, -21.956466464) # color=#2EE6D6 width=2
```
but it should be
```
# Region file format: DS9 CARTA 6.0.0-dev
global color=green dashlist=8 3 width=1 font="helvetica 10 normal roman" select=1 highlite=1 dash=0 fixed=0 edit=1 move=1 delete=1 include=1 source=1
fk5
point(269.090232042, -21.955048017) # color=#2EE6D6 width=2
line(269.087708606, -21.955757242, 269.089696784, -21.957175681) # color=#2EE6D6 width=2
box(269.088702690, -21.956466464, 6.1277", 8.1702", 0) # color=#2EE6D6 width=2
ellipse(269.088702690, -21.956466464, 2.5532", 5.1064", 90) # color=#2EE6D6 width=2
polygon(269.087555672, -21.955402630, 269.087937997, -21.958168590, 269.088702690, -21.956466464) # color=#2EE6D6 width=2
polyline(269.088702690, -21.957884904, 269.089926194, -21.957601211, 269.087479189, -21.957317524, 269.088702690, -21.956466464) # color=#2EE6D6 width=2
```

2) if we define a ds9 vector annotation like

```
# vector(269.0886872,-21.9537391,4.533",276.96825) vector=1
```
after import, in the region list widget we see its center coordinate is bad
<img width="664" height="65" alt="Screenshot 2025-09-22 at 14 43 25" src="https://github.com/user-attachments/assets/557d999c-955e-4d8c-8696-40e371e6227a" />

3) if we import a ds9 region in image coordinates like

```
# Region file format: DS9 version 4.1
global color=green dashlist=8 3 width=4 font="helvetica 10 normal roman" select=1 highlite=1 dash=0 fixed=0 edit=1 move=1 delete=1 include=1 source=1
image
circle(114.99968,348.00297,20)
ellipse(99.00041,269.99799,18,49,360)
box(148.00311,140.00244,98,54,360)
polygon(90.000834,98.997886,111.00254,68.002125,130.00029,98.998304,130.00063,59.002304,116.99945,32.002185,90.00133,59.001886)
line(280.00324,365.00248,360.99802,277.99686) # line=0 0
# vector(216.00244,394.99787,90.66,276.96826) vector=1
# projection(261.99923,261.99937,299.00043,193.00155,31.68)
# segment(171.00056,286.99777,183.00081,230.00262,235.00071,194.00266)
# text(288.00182,146.00003) text={carta}
point(289.99796,75.000813) # point=circle 5
point(346.99904,99.998648) # point=cross 5
point(289.99781,53.400813) # point=circle
point(346.99877,78.398648) # point=cross
# ruler(239.00053,116.99865,205.99932,49.001868) ruler=fk5 degrees
```
The last one "ruler" is missing after import.

4) when exporting regions in ds9 format, we see the coordinate reference changes from fk5 to J2000

```
# compass(294.63, 290.54, 100.00) color=#FFBA01 width=2 font="helvetica 20 normal roman" compass=J2000 {N} {E} 1 1
# ruler(214.97, 198.63, 123.05, 152.67) color=#FFBA01 width=2 ruler=J2000 degrees
```

pford committed:
Fix DS9 export/import and add tests for imported==original region

pford commented:
@kswang1029 I think I fixed the issues, and added a check to the backend tests that the exported-then-imported regions are the same as the original.  These tests check all region types in CRTF/DS9 pixel and world coordinates.

kswang1029 reviewed (changes requested):
@pford there is still one issue:

With a ds9 region in image coordinates
```
# Region file format: DS9 version 4.1
global color=green dashlist=8 3 width=4 font="helvetica 10 normal roman" select=1 highlite=1 dash=0 fixed=0 edit=1 move=1 delete=1 include=1 source=1
image
# ruler(239.00053,116.99865,205.99932,49.001868) ruler=fk5 degrees
# compass(379.00287,368.99739,40) compass=fk5 {N} {E} 1 1
```

Ruler and compass are not imported successfully thus are missing in the region list widget with the error messages in the terminal

```
Invalid DS9 region syntax: rulerFailed to apply compass to image.
Invalid DS9 region syntax: compass
```

pford commented:
> @pford there is still one issue:
> 
> With a ds9 region in image coordinates
> 
> ```
> # Region file format: DS9 version 4.1
> global color=green dashlist=8 3 width=4 font="helvetica 10 normal roman" select=1 highlite=1 dash=0 fixed=0 edit=1 move=1 delete=1 include=1 source=1
> image
> # ruler(239.00053,116.99865,205.99932,49.001868) ruler=fk5 degrees
> # compass(379.00287,368.99739,40) compass=fk5 {N} {E} 1 1
> ```
> 
@kswang1029 I think `ruler=fk5 degrees` and `compass=fk5` describe these region world coordinates and override the global coordinate system `image` (pixels)?  Else if pixel coordinates it would be `ruler=image pixels` and `compass=image`?

I exported ruler and compass regions in pixel and world coordinates to DS9 region files, then pasted the "world" ruler and compass lines into the pixel region file with global `image`.  These regions were imported correctly as interpreted to be fk5 degrees.

**Edit:** I am trying to import these files into the SAOImage ds9 application and they are clearly wrong!  Will look into it.

pford committed:
Fix DS9 import and export

pford commented:
@kswang1029 I was mistaken, the ruler and compass coordinate frame are for the image (world coordinate frame) even for image pixels import/export.  I have tested this in CARTA and in the DS9 image viewer and it should be fixed now.

kswang1029 reviewed (approved):
looks good. 👍
